futures-util/src/async_stream/await_item.rs

```diff
@@ -0,0 +1,46 @@
+/// Await an item from a stream inside an `async` function.
+///
+/// You should pass an object implementing [`Stream`] to this macro, it will
+/// implicitly `yield` while that stream returns [`Poll::Pending`] and evaluate
+/// to a [`Result`] containing the next item or error from that stream.
+///
+/// If you want to iterate over all items in a `Stream` you should instead see
+/// the documentation on `#[async] for` in the main `#[async]` documentation.
+///
+/// # Examples
+///
+/// ```no-run
+/// #![feature(futures_api, async_await, generators)]
+/// use futures::prelude::*;
+/// use futures::executor::block_on;
+/// use futures::await_item;
+///
+/// async fn eventually_ten() -> u32 {
+///     let mut stream = stream::repeat(5);
+///     if let Some(first) = await_item!(&mut stream) {
+///         if let Some(second) = await_item!(&mut stream) {
+///             return first + second;
+///         }
+///     }
+///     0
+/// }
+///
+/// assert_eq!(10, block_on(eventually_ten()));
+/// ```
+#[macro_export]
+macro_rules! await_item {
```

Comment:
Is this necessary now that we have `await!(stream.next())`?

Reply:
I was concerned about `Self: Unpin`, but https://github.com/rust-lang-nursery/futures-rs/issues/1547#issuecomment-486202839 would be a better solution. So I will remove this macro.

Reply:
Within the context of a `static`-generator stack-pinning works just like normal. Currently all `static` generators are `!Unpin`, but if they could be automatically made `Unpin` if they didn't have self-references, then the self-reference used in `Next` would make sure they are correctly `Unpin` or not.